internal/liveness/metrics.go

```diff
@@ -0,0 +1,135 @@
+	client := util.NewK8sClient()
+	pvcs, err := client.CoreV1().PersistentVolumeClaims("").List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		klog.Errorf("Get pvclist failed: %v", err)
+		return
+	}
+	for _, pvc := range pvcs.Items {
+		pvcName := pvc.Name
+		pvcNameSpace := pvc.Namespace
+		pvName := pvc.Spec.VolumeName
+		pvInfo, err := client.CoreV1().PersistentVolumes().Get(context.TODO(), pvName, metav1.GetOptions{})
+		if err != nil {
+			klog.V(4).Infof("Get pvinfo failed: %v", err)
+			continue
+		}
+		if pvInfo.Spec.CSI != nil {
+			volumeHandle := pvInfo.Spec.CSI.VolumeHandle
+			volumeHandleSlice := strings.Split(volumeHandle, "-")
+			index := len(volumeHandleSlice) - 5
+			imageName := strings.Join(append([]string{"csi-vol"}, volumeHandleSlice[index:]...), "-")
+			var secretName string
+			var secretNamespace string
+			if pvInfo.Spec.CSI.ControllerExpandSecretRef != nil {
+				secretName = pvInfo.Spec.CSI.ControllerExpandSecretRef.Name
+				secretNamespace = pvInfo.Spec.CSI.ControllerExpandSecretRef.Namespace
+			} else {
+				secretName = pvInfo.Spec.CSI.NodePublishSecretRef.Name
+				secretNamespace = pvInfo.Spec.CSI.NodePublishSecretRef.Namespace
+			}
+			imageCluster := pvInfo.Spec.CSI.VolumeAttributes["clusterID"]
+			monitors, err := util.Mons(csiConfigFile, imageCluster)
+			if err != nil {
+				klog.Error(err.Error())
+				return
+			}
+			secret, _ := client.CoreV1().Secrets(secretNamespace).Get(context.TODO(), secretName, metav1.GetOptions{})
+			newSecrets := make(map[string]string)
+			for key, value := range secret.Data {
+				newSecrets[key] = string(value)
+			}
+			cr, err := util.NewUserCredentials(newSecrets)
+			if err != nil {
+				klog.Error(err.Error())
+				return
+			}
+			imagePool := pvInfo.Spec.CSI.VolumeAttributes["pool"]
+			volumeName := pvInfo.Spec.CSI.VolumeAttributes["fsName"]
+			if imagePool == "" && volumeName == "" {
+				//unknown volume
+				continue
+			} else if volumeName != "" {
+				//cephfs volume
+				subvolumeGroup, err := util.CephFSSubvolumeGroup(csiConfigFile, imageCluster)
+				if err != nil {
+					klog.Errorf(err.Error())
+					subvolumeGroup = "csi"
+				}
+				if CheckSubVolumeInCluster(monitors, imageName, subvolumeGroup, volumeName, cr) {
+					cephfsMapping.WithLabelValues(pvcName, pvcNameSpace, pvName, volumeName, subvolumeGroup, imageName).Set(1)
+				} else {
+					cephfsMapping.WithLabelValues(pvcName, pvcNameSpace, pvName, volumeName, subvolumeGroup, imageName).Set(0)
+				}
+			} else {
+				//rbd volume
+				if CheckImageInCluster(imageName, imagePool, monitors, cr) {
+					rbdMapping.WithLabelValues(pvcName, pvcNameSpace, pvName, imagePool, imageName).Set(1)
+				} else {
+					rbdMapping.WithLabelValues(pvcName, pvcNameSpace, pvName, imagePool, imageName).Set(0)
+				}
+			}
+		}
+	}
+}
+
+func recordMapping(pollTime, timeout time.Duration) {
+	// register prometheus metrics
+	err := prometheus.Register(rbdMapping)
```

Comment:
we should not handle both cephfs and rbd in one liveness

Reply:
done